### ISSUE TITLE
Same dependencies with different classifiers are not consumed correctly from Gradle Module Metadata

### ISSUE THREAD

**Issue by mindstorm38** [gradle/gradle#11825]
I think the `.module` file produced by `maven-publish` plugin (next to POM) bring Gradle to ignore redundant dependencies artifacts (for example if we want in compileClasspath two archives of one artifact, using classifiers to select them) in each `dependencies` section of each `variants` in the JSON-like module file when "consuming" a maven artifact.

### Expected Behavior
For example if I've a project A, a library and I want it to depend on _(btw I'm using Kotlin DSL)_ :
- `api("com.example:foo:1.0.0")`
- `runtimeOnly("com.example:foo:1.0.0:natives-windows")` (inspired by lwjgl 3 libs if you know)
Then I publish this library A on a maven repo _(or local maven repo, same behavior)_ and with same group and version 1.0.0.

Next I create a project B that depends on A :
- `implementation("com.example:A:1.0.0")`

Finally, I expected Gradle to "apply transitivity"  _(I don't really know if it's the real name of this behavior)_, that is, Gradle to append in the project B `compileClasspath` the `com.example:foo:1.0.0` artifact, and in the `runtimeClasspath` the natives' artifact `com.example:foo:1.0.0:natives-windows`.

### Current Behavior
But in reality Gradle does only one of them, adding natives artifact to runtimeClasspath, or default artifact to compileClasspath. It's the same with more artifacts or more classifiers.

### Context
I'm making a 3D engine helper onto LWJGL 3, and wanted to include to this library all three natives in addition to each needed default artifacts of LWJGL, all that published on Maven Central for use in my Minecraft clone.

### Steps to Reproduce 
I think I've described how to reproduce in top sections, but if you want to test with my projects :
- [msengine](https://gitlab.com/mindstorm38/msengine/tree/9c1cd0fe8174280536a8725d333bdabe96643234), just by launching task `publishToMavenLocal`
- [proc-gen](https://github.com/mindstorm38/proc-gen/tree/7b9505148df634f8021bb34d12af494ff474e1d8), by trying `showConf` tasks that display `runtimeClasspath` (which I'm interested in for natives) configuration artifacts archives.

**Finally, I solved this problem by deleting the module file from my local maven repo, and showConf task worked as expected. For now the only way I can get rid of this issue is to use only maven POM file.**

### Your Environment
Build scan URL: [scan (page not found for me...)](https://scans.gradle.com/s/he7zs4r3wglf2s)

**Comment by mindstorm38** [comment]
I found a cleaner way to solve this issue using the `metadataSources` blocks of MavenArtifactRepository class, for example here I configure this for the mavenLocal repository :
```kotlin
repositories {
    mavenLocal {
        metadataSources {
            mavenPom()
            ignoreGradleMetadataRedirection()
        }
    }
}
```

**Comment by ljacomet** [comment]
Thanks for the report, this is indeed a bug that will need to be fixed.
I managed to reproduce it in isolation.

First of all, the produced Gradle Module Metadata _is correct_. However, the consumption is not.
So the workaround of ignoring Gradle module metadata redirection is the correct one for now.

What is currently not working is resolving a single variant in Gradle Module Metadata that declares a dependency on the same module, once with a classifier, once normal. Gradle loses the one without classifier information.

**Comment by mindstorm38** [comment]
Thanks for information, that is what I thought about the resolving.

**Comment by jjohannes** [comment]
Fixed for Gradle 6.3

**Comment by jjohannes** [comment]
We will do another 6.3 release candidate. We already have a nightly with the fix. Would you care to confirm that it solves the issue for your @mindstorm38?

`./gradlew wrapper --gradle-version=6.3-20200318000026+0000`

**Comment by mindstorm38** [comment]
**_It works_**

**Comment by jjohannes** [comment]
We just released 6.3 RC3 with this fix included.
